### ISSUE TITLE
Your pretrained model is still wrong.

### ISSUE THREAD

**Issue by HospitableHost** [garvita-tiwari/PoseNDF#11]
When I use your pretrained model to calculate the distance of the poses from AMASS, I get 'nan'.
For example, the pose (already in quaternion, and it is sampled from AMASS with your code):
       [[ 0.9198, -0.3837, -0.0681,  0.0455],
        [ 0.9138, -0.4025,  0.0352, -0.0405],
        [ 0.9497,  0.3132,  0.0031, -0.0044],
        [ 0.7995,  0.5880,  0.0242, -0.1203],
        [ 0.7601,  0.6438, -0.0749,  0.0462],
        [ 0.9952, -0.0918,  0.0040,  0.0338],
        [ 0.9120, -0.3939,  0.0701,  0.0904],
        [ 0.9162, -0.3983, -0.0442, -0.0026],
        [ 1.0000,  0.0054,  0.0021,  0.0038],
        [ 1.0000,  0.0000,  0.0000,  0.0000],
        [ 1.0000,  0.0000,  0.0000,  0.0000],
        [ 0.9611, -0.2573,  0.0985,  0.0199],
        [ 0.9835, -0.0572, -0.0382, -0.1676],
        [ 0.9785, -0.0461,  0.0593,  0.1920],
        [ 0.9969,  0.0629, -0.0383, -0.0263],
        [ 0.9057,  0.0084, -0.1829, -0.3823],
        [ 0.9148,  0.0618,  0.1531,  0.3687],
        [ 0.5447, -0.0337, -0.7993,  0.2514],
        [ 0.5862, -0.0814,  0.7476, -0.3014],
        [ 0.9316, -0.2648, -0.1939,  0.1562],
        [ 0.9618, -0.2100,  0.1656, -0.0593]])
its distance is 'nan', when using your pretrained model.

**Comment by garvita-tiwari** [comment]
I tried with this value of pose and I am getting 0.0000087.
Are you getting nans for other poses also ?

**Comment by HospitableHost** [comment]
> I tried with this value of pose and I am getting 0.0000087. Are you getting nans for other poses also ?

yes. Besides, I used the yaml file you updated yesterday, and the result is still nan for all poses.

**Comment by HospitableHost** [comment]
> I tried with this value of pose and I am getting 0.0000087. Are you getting nans for other poses also ?

my test code is below:
![image](https://user-images.githubusercontent.com/44576391/215275684-156396ef-d82d-411f-bc81-10f601c9dd41.png)
the result is below:
![image](https://user-images.githubusercontent.com/44576391/215275701-10ffde71-2b04-4c2b-9307-94faa71a39ef.png)

**Comment by HospitableHost** [comment]
> I tried with this value of pose and I am getting 0.0000087. Are you getting nans for other poses also ?

more intuitively：
![image](https://user-images.githubusercontent.com/44576391/215675291-87b23fe1-3d70-4b0d-9080-6463717bef1e.png)
![image](https://user-images.githubusercontent.com/44576391/215675344-d5a67f91-9fd9-46e3-b14e-fd8f7bd17baa.png)

**Comment by twehrbein** [comment]
I am also getting NANs with the pretrained model. They appear after the 4th layer of DFNet.
Would be great if you could fix that asap.

**Comment by raypine** [comment]
It can be observed that ``model.dfnet.lin2.weight_v[62:125]`` are all zeros, and this will cause weight normalization to produre ``nan`` gradually. Is this the reason ? @garvita-tiwari

**Comment by garvita-tiwari** [comment]
> I am also getting NANs with the pretrained model. They appear after the 4th layer of DFNet. Would be great if you could fix that asap.

 I checked for my case, I don't have nans. Please share the pytorch3d version you are using

**Comment by HospitableHost** [comment]
> > I am also getting NANs with the pretrained model. They appear after the 4th layer of DFNet. Would be great if you could fix that asap.
> 
> I checked for my case, I don't have nans. Please share the pytorch3d version you are using

0.7.2

**Comment by HospitableHost** [comment]
> > I am also getting NANs with the pretrained model. They appear after the 4th layer of DFNet. Would be great if you could fix that asap.
> 
> I checked for my case, I don't have nans. Please share the pytorch3d version you are using

The code in PoseNDF class & DFNet class & BoneMLP class don't include pytorch3d, I think it has nothing to do with pytorch3d.
Besides, I input the tensor directly to the PoseNDF model.
![image](https://user-images.githubusercontent.com/44576391/216511181-48d9d920-c615-4075-b444-20efeb00fea6.png)

**Comment by twehrbein** [comment]
> > I am also getting NANs with the pretrained model. They appear after the 4th layer of DFNet. Would be great if you could fix that asap.
> 
> I checked for my case, I don't have nans. Please share the pytorch3d version you are using

I am using pytorch3d 0.7.0, but I do not believe this to be the problem.
Like @raypine said, model.dfnet.lin2.weight_v[62:125] are all zeros which results in the corresponding vector norms to be 0 and thus producing nans at division. Could you check the values for model.dfnet.lin2.weight_v[62:125] at your end?

**Comment by garvita-tiwari** [comment]
I wasn't able to figure out the reason, because I am not getting nans. But I have uploaded another trained model here.
 [model-test](https://nextcloud.mpi-klsb.mpg.de/index.php/s/ftHC7ZLdgpYQtB8). 

This is not the converged model, please try with this model

**Comment by HospitableHost** [comment]
> I wasn't able to figure out the reason, because I am not getting nans. But I have uploaded another trained model here. [model-test](https://nextcloud.mpi-klsb.mpg.de/index.php/s/ftHC7ZLdgpYQtB8).
> 
> This is not the converged model, please try with this model

The new one works, and the predicted distance is 0.
But it seems that you uploaded the older one than this new one before.
this new one's(37 minutes ago) epoch is 10990 epoch, and the old one's(19 days ago) epoch is 2058.

**Comment by HospitableHost** [comment]
> I wasn't able to figure out the reason, because I am not getting nans. But I have uploaded another trained model here. [model-test](https://nextcloud.mpi-klsb.mpg.de/index.php/s/ftHC7ZLdgpYQtB8).
> 
> This is not the converged model, please try with this model

I find that this new model output 0 for any pose, even for the noised pose.

**Comment by twehrbein** [comment]
> > I wasn't able to figure out the reason, because I am not getting nans. But I have uploaded another trained model here. [model-test](https://nextcloud.mpi-klsb.mpg.de/index.php/s/ftHC7ZLdgpYQtB8).
> > This is not the converged model, please try with this model
> 
> I find that this new model output 0 for any pose, even for the noised pose.

Can confirm...
@garvita-tiwari are you sure that the uploaded code is correct? E.g. in net_modules.py, ReLU is always used, no matter what argument is given for "activation".

**Comment by ktxu1224** [comment]
@garvita-tiwari  I have the same problem of NaN, can you check it? For example, when input the array as follows:
tensor([[ 0.0699, -0.0831, -0.0769,  0.9911],
        [-0.2234,  0.0264, -0.0082,  0.9743],
        [ 0.0828, -0.0686, -0.0169,  0.9941],
        [ 0.0846, -0.2430, -0.0217,  0.9661],
        [ 0.4965, -0.0064,  0.1000,  0.8622],
        [ 0.1408, -0.0339,  0.0597,  0.9877],
        [-0.1659,  0.0399,  0.0991,  0.9803],
        [-0.1475, -0.0170,  0.0091,  0.9889],
        [-0.0660, -0.0240, -0.0311,  0.9970],
        [-0.0039, -0.0094, -0.0050,  0.9999],
        [ 0.0088,  0.0156,  0.0063,  0.9998],
        [ 0.0247, -0.1040,  0.0048,  0.9943],
        [ 0.0776,  0.0517, -0.2211,  0.9708],
        [ 0.0370, -0.0209,  0.2253,  0.9734],
        [-0.0177, -0.0214,  0.1001,  0.9946],
        [ 0.1208, -0.1519, -0.4285,  0.8824],
        [ 0.1129,  0.0452,  0.5209,  0.8449],
        [ 0.0569, -0.6230,  0.0448,  0.7789],
        [ 0.0050,  0.1621, -0.1155,  0.9800],
        [-0.1123, -0.0927, -0.2624,  0.9539],
        [-0.1226,  0.0099,  0.2475,  0.9610]], device='cuda:0',
       grad_fn=<SelectBackward0>)

**Comment by ktxu1224** [comment]
@HospitableHost hi, have you figure out the problem of NaN?

**Comment by HospitableHost** [comment]
> @HospitableHost hi, have you figure out the problem of NaN?

No, the model is wrong, so nobody can fix it.

**Comment by ktxu1224** [comment]
> > @HospitableHost hi, have you figure out the problem of NaN?
> 
> No, the model is wrong, so nobody can fix it.

@garvita-tiwari  hi，i think posendf makes sense and can be used to constrain the reasonable space of the motion of animation,  but there are some small error in current uploaded code or model, so can you upload correct code and model for testing? Let's improve the accuracy and generalization of the model, thanks a lot

**Comment by garvita-tiwari** [comment]
Hi, I have figured out a bug in data and now training the model again.  I will share the model asap(within 1-2 days) if there are no other bugs.

**Comment by TimofeyKruk** [comment]
> Hi, I have figured out a bug in data and now training the model again. I will share the model asap(within 1-2 days) if there are no other bugs.

@garvita-tiwari  
Thank you!!!

**Comment by HospitableHost** [comment]
> Hi, I have figured out a bug in data and now training the model again. I will share the model asap(within 1-2 days) if there are no other bugs.

Hi, I find that you uploaded a new one yesterday, and then I tested it. And it seems still not good.
I have tested it using three activation functions(relu, lrelu and softplus), and it outputs the same order of magnitude for plausible poses and noisy poses.
(same for two kinds of noisy poses: sampled_pose = sampled_pose + sigma*np.random.rand(21,4)*sampled_pose
                                                  and: sampled_pose = sampled_pose + sigma*np.random.rand(21,4))

**Comment by garvita-tiwari** [comment]
> > Hi, I have figured out a bug in data and now training the model again. I will share the model asap(within 1-2 days) if there are no other bugs.
> 
> Hi, I find that you uploaded a new one yesterday, and then I tested it. And it seems still not good. I have tested it using three activation functions(relu, lrelu and softplus), and it outputs the same order of magnitude for plausible poses and noisy poses. (same for two kinds of noisy poses: sampled_pose = sampled_pose + sigma*np.random.rand(21,4)_sampled_pose and: sampled_pose = sampled_pose + sigma_np.random.rand(21,4))

Hi yes this is not the optimal mode, as it was trained on a subset of AMASS. I will upload the optimal model soon

**Comment by ktxu1224** [comment]
> > > Hi, I have figured out a bug in data and now training the model again. I will share the model asap(within 1-2 days) if there are no other bugs.
> > 
> > 
> > Hi, I find that you uploaded a new one yesterday, and then I tested it. And it seems still not good. I have tested it using three activation functions(relu, lrelu and softplus), and it outputs the same order of magnitude for plausible poses and noisy poses. (same for two kinds of noisy poses: sampled_pose = sampled_pose + sigma*np.random.rand(21,4)_sampled_pose and: sampled_pose = sampled_pose + sigma_np.random.rand(21,4))
> 
> Hi yes this is not the optimal mode, as it was trained on a subset of AMASS. I will upload the optimal model soon

@garvita-tiwari hi, whether the optimal model is ready?

**Comment by garvita-tiwari** [comment]
Hi,

Please check version2 branch and model(and config) here:
https://nextcloud.mpi-klsb.mpg.de/index.php/s/dFbra3gfycYc5mz

In the original model, we applied normalization along the wrong axis:
pose = torch.nn.functional.normalize(pose.to(device=self.device),dim=1)

Although experimentally, it doesn't degrade the performance. We suspect it's because during training all the input quaternions are already unit quaternion and this particular layer act as a normalization layer. 

 
This should be:
pose = torch.nn.functional.normalize(pose.to(device=self.device),dim=-1)
In version2, we have moved data normalization step and

**Comment by HospitableHost** [comment]
> Hi,
> 
> Please check version2 branch and model(and config) here: https://nextcloud.mpi-klsb.mpg.de/index.php/s/dFbra3gfycYc5mz
> 
> In the original model, we applied normalization along the wrong axis: pose = torch.nn.functional.normalize(pose.to(device=self.device),dim=1)
> 
> Although experimentally, it doesn't degrade the performance. We suspect it's because during training all the input quaternions are already unit quaternion and this particular layer act as a normalization layer.
> 
> This should be: pose = torch.nn.functional.normalize(pose.to(device=self.device),dim=-1) In version2, we have moved data normalization step and

the output poses from the function 'axis_angle_to_quaternion' is already normalized, I think. So, it seems that we don't need the 'pose = torch.nn.functional.normalize(pose.to(device=self.device),dim=-1)'.

**Comment by garvita-tiwari** [comment]
> > Hi,
> > Please check version2 branch and model(and config) here: https://nextcloud.mpi-klsb.mpg.de/index.php/s/dFbra3gfycYc5mz
> > In the original model, we applied normalization along the wrong axis: pose = torch.nn.functional.normalize(pose.to(device=self.device),dim=1)
> > Although experimentally, it doesn't degrade the performance. We suspect it's because during training all the input quaternions are already unit quaternion and this particular layer act as a normalization layer.
> > This should be: pose = torch.nn.functional.normalize(pose.to(device=self.device),dim=-1) In version2, we have moved data normalization step and
> 
> the output poses from the function 'axis_angle_to_quaternion' is already normalized, I think. So, it seems that we don't need the 'pose = torch.nn.functional.normalize(pose.to(device=self.device),dim=-1)'.

Yes the output is already normalized. You don't need this step. I am keeping this, in case I don't have axis_angle_to_quaternion performed in previous step. e.g. in experiments/sample_poses.py
